master/internal/stream/metrics.go

```diff
@@ -198,99 +199,37 @@ func MetricCollectStartupMsgs(
 	return out, nil
 }
 
-// MetricCollectSubscriptionModMsgs scrapes the database when a
-// user submits a new MetricSubscriptionSpec for initial matches.
-func MetricCollectSubscriptionModMsgs(ctx context.Context, addSpec MetricSubscriptionSpec) (
-	[]interface{}, error,
-) {
-	if len(addSpec.MetricIds) == 0 && len(addSpec.TrialIds) == 0 && len(addSpec.ExperimentIds) == 0 {
-		return nil, nil
-	}
-	var metricMsgs []*MetricMsg
-	q := getMetricMsgsWithWorkspaceID(metricMsgs)
-
-	// Use WhereSince to build a complex WHERE clause.
-	ws := stream.WhereSince{Since: addSpec.Since}
-	if len(addSpec.MetricIds) > 0 {
-		ws.Include("id in (?)", bun.In(addSpec.MetricIds))
-	}
-	q = ws.Apply(q)
-
-	err := q.Scan(ctx)
-	if err != nil && errors.Cause(err) != sql.ErrNoRows {
-		log.Errorf("error: %v\n", err)
-		return nil, err
-	}
-
-	var out []interface{}
-	for _, msg := range metricMsgs {
-		out = append(out, msg.UpsertMsg())
-	}
-	return out, nil
-}
-
-// MetricFilterMaker tracks the metric id's that are to be filtered for.
-type MetricFilterMaker struct {
-	MetricIds     map[int]bool
-	TrialIds      map[int]bool
-	ExperimentIds map[int]bool
-}
-
-// NewMetricFilterMaker creates a new FilterMaker.
-func NewMetricFilterMaker() FilterMaker[*MetricMsg, MetricSubscriptionSpec] {
-	return &MetricFilterMaker{
-		make(map[int]bool),
-		make(map[int]bool),
-		make(map[int]bool),
-	}
-}
-
-// AddSpec adds MetricIds specified in MetricSubscriptionSpec.
-func (ms *MetricFilterMaker) AddSpec(spec MetricSubscriptionSpec) {
-	for _, id := range spec.MetricIds {
-		ms.MetricIds[id] = true
-	}
-	for _, id := range spec.TrialIds {
-		ms.TrialIds[id] = true
-	}
-	for _, id := range spec.ExperimentIds {
-		ms.ExperimentIds[id] = true
+// MetricMakeFilter creates a MetricMsg filter based on the given MetricSubscriptionSpec.
+func MetricMakeFilter(spec *MetricSubscriptionSpec) (func(*MetricMsg) bool, error) {
+	// Should this filter even run?
+	if len(spec.MetricIds) == 0 && len(spec.TrialIds) == 0 && len(spec.ExperimentIds) == 0 {
+		return nil, fmt.Errorf(
+			"invalid subscription spec arguments: %v %v %v",
+			spec.MetricIds, spec.TrialIds, spec.ExperimentIds,
+		)
 	}
-}
 
-// DropSpec removes MetricIds specified in MetricSubscriptionSpec.
-func (ms *MetricFilterMaker) DropSpec(spec MetricSubscriptionSpec) {
+	// Make a copy of the map, because the filter must run safely off-thread.
+	metricIds := make(map[int]struct{})
+	trialIds := make(map[int]struct{})
+	experimentIds := make(map[int]struct{})
```

Comment:
can we not expose per-experiment metrics for now?  Or, how much does it cost us performance-wise to filter by experiment?

If it has even a moderate cost, I'd almost prefer to be drop support for this, and have users subscribe to every trial's metrics in an experiment in the case that they want every experiment metric.

But if we already have to join the experiments table for something else then maybe it makes no difference.

Reply:
We're already doing a join on the experiments table to get to workspace information for permission checks, so the only additional cost is passing the additional value through the channel and underlying trigger functions